sys/linux/socket_ip_tunnel.txt

```diff
@@ -0,0 +1,93 @@
+iphdr {
+	ihl		bytesize4[parent, int8:4]
+	version		const[4, int8:4]
+	ecn		int8:2
+	dscp		int8:6
+	total_len	bytesize4[parent, int16be]
+	id		int16be[100:104]
+	frag_off	flags[frag_off_flags, int16be]
+	ttl		int8
+	protocol	flags[ip_tunnel_protocols, int8]
+	csum		csum[parent, inet, int16be]
+	src_ip		ipv4_addr
+	dst_ip		ipv4_addr
+} [packed]
+
+# IPv4 tunnel parameters
+ip_tunnel_parm {
+	name	string[ipv4_tunnel_names, IFNAMSIZ]
+	link	int32
+	i_flags	flags[ip_tunnel_flags, int16be]
+	o_flags	flags[ip_tunnel_flags, int16be]
+	i_key	int32be
+	o_key	int32be
+	iph	iphdr
+}
+
+# IPv6 tunnel parameters
+ip6_tnl_parm2 {
+	name		string[ipv6_tunnel_names, IFNAMSIZ]
+	link		int32
+	proto		int8
+	encap_limit	int8
+	hop_limit	int8
+	flowinfo	int32be
+	flags		int32
+	laddr		ipv6_addr
+	raddr		ipv6_addr
+	i_flags		flags[ip_tunnel_flags, int16be]
+	o_flags		flags[ip_tunnel_flags, int16be]
+	i_key		int32be
+	o_key		int32be
+}
+
+type ifreq_ipv4_tunnel ifreq_dev_t[ipv4_tunnel_names, ptr[inout, ip_tunnel_parm]]
+type ifreq_ipv6_tunnel ifreq_dev_t[ipv6_tunnel_names, ptr[inout, ip6_tnl_parm2]]
+
+# IPv4 ioctls
+ioctl$sock_ipv4_tunnel_SIOCGETTUNNEL(fd sock_ipv4, cmd const[SIOCGETTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+ioctl$sock_ipv4_tunnel_SIOCADDTUNNEL(fd sock_ipv4, cmd const[SIOCADDTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+ioctl$sock_ipv4_tunnel_SIOCDELTUNNEL(fd sock_ipv4, cmd const[SIOCDELTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+ioctl$sock_ipv4_tunnel_SIOCCHGTUNNEL(fd sock_ipv4, cmd const[SIOCCHGTUNNEL], arg ptr[inout, ifreq_ipv4_tunnel])
+
+# IPv6 ioctls
+# SIOC***TUNNEL
+ioctl$sock_ipv6_tunnel_SIOCGETTUNNEL(fd sock_ipv6, cmd const[SIOCGETTUNNEL], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCADDTUNNEL(fd sock_ipv6, cmd const[SIOCADDTUNNEL], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCDELTUNNEL(fd sock_ipv6, cmd const[SIOCDELTUNNEL], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCCHGTUNNEL(fd sock_ipv6, cmd const[SIOCCHGTUNNEL], arg ptr[inout, ifreq_ipv6_tunnel])
+# SIOC***PRL
+ioctl$sock_ipv6_tunnel_SIOCGETPRL(fd sock_ipv6, cmd const[SIOCGETPRL], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCADDPRL(fd sock_ipv6, cmd const[SIOCADDPRL], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCDELPRL(fd sock_ipv6, cmd const[SIOCDELPRL], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCCHGPRL(fd sock_ipv6, cmd const[SIOCCHGPRL], arg ptr[inout, ifreq_ipv6_tunnel])
+# SIOC***6RD
+ioctl$sock_ipv6_tunnel_SIOCGET6RD(fd sock_ipv6, cmd const[SIOCGET6RD], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCADD6RD(fd sock_ipv6, cmd const[SIOCADD6RD], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCDEL6RD(fd sock_ipv6, cmd const[SIOCDEL6RD], arg ptr[inout, ifreq_ipv6_tunnel])
+ioctl$sock_ipv6_tunnel_SIOCCHG6RD(fd sock_ipv6, cmd const[SIOCCHG6RD], arg ptr[inout, ifreq_ipv6_tunnel])
```

Comment:
Since the ioctls are pretty repetitive, we can probably simplify them using flags for `cmd` instead of const.

Reply:
From [here](https://github.com/google/syzkaller/issues/1913#issuecomment-655659067):

> We try to not do this lately, though, for the following reasons.
> flags is generally considered more "permissive" wrt what values can be passed.
> These ioctl variants are really different syscalls for most practical purposes, so we tried to stick to convention: one kernel > syscall -> one syscall variant in the descriptions. This may allow better heuristics in future, e.g. what should be a priority > of one syscall as compared to another. Blending effectively several syscalls into a single variant in descriptions makes such reasoning harder.
> But maybe we need something else to support these patterns better.

Reply:
Okay, got it. Based on it I assume I should stick with the "one syscall, one variant" pattern for now.